components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APIAuthenticationHandler.java

```diff
@@ -77,13 +79,22 @@
 public class APIAuthenticationHandler extends AbstractHandler implements ManagedLifecycle {
     private static final Log log = LogFactory.getLog(APIAuthenticationHandler.class);
 
-    private volatile Authenticator authenticator;
+    private volatile ArrayList<Authenticator> authenticators = null;
```

Comment:
I think we should remove this volatile, I dont see a reason why we should have it and it can lead to performance issues

Reply:
Updated